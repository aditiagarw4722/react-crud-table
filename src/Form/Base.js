// @flow
import React from 'react';
import { Formik, Field } from 'formik';

import { Form } from './wrappers';
import Button from '../Button';

import { DEFAULT_VALIDATE, generalValidationError } from './helpers';

type FieldData = {
  name: string,
  label: string,
  placeholder?: string,
  render?: Function | string,
  type?: string,
  readOnly?: boolean,
};

type FormBaseProps = {
  touched: Object,
  errors: Object,
  data: {
    message?: string,
    generalErrorMessage?: string,
    fields: Array<FieldData>,
    submitButtonProps: Object,
    submitText: string,
  },
  error: ?string,
};

const FormBase = ({
  data,
  errors,
  touched,
  error,
}: FormBaseProps): React$Element<any> => (
  <Form>
    {data.message && <Form.Message>{data.message}</Form.Message>}
    {error && <Form.ErrorMessage>{error}</Form.ErrorMessage>}
    {generalValidationError({
      touched,
      errors,
      message: data.generalErrorMessage,
    })}
    {data.fields.map((field) => (
      <Form.FieldContainer key={field.name}>
        <Form.Label htmlFor={field.name}>{field.label}</Form.Label>
        <Field
          name={field.name}
          placeholder={field.placeholder}
          render={field.render}
          type={field.type}
          readOnly={field.readOnly}
        />
        {errors[field.name] && touched[field.name] && (
          <Form.FieldError>{errors[field.name]}</Form.FieldError>
        )}
      </Form.FieldContainer>
    ))}
    {generalValidationError({
      touched,
      errors,
      message: data.generalErrorMessage,
    })}
    <Button type="submit" modifiers="positive" {...data.submitButtonProps}>
      {data.submitText}
    </Button>
  </Form>
);

export default FormBase;
