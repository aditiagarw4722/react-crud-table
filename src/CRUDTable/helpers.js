import React from 'react';
import {
  SORT_DIRECTIONS,
  FIELDS_COMPONENT_TYPE,
  FIELD_COMPONENT_TYPE,
  CREATE_FORM_COMPONENT_TYPE,
  DELETE_FORM_COMPONENT_TYPE,
  UPDATE_FORM_COMPONENT_TYPE,
  PAGINATION_COMPONENT_TYPE,
} from './constants';

export const chevron = (direction) => (direction === SORT_DIRECTIONS.ASCENDING ? (
  <span>&#x25B2;</span>
) : (
  <span>&#x25BC;</span>
));

export const toggleDirection = (direction, toggle) => {
  if (toggle) {
    switch (direction) {
      case 'ascending':
        return 'descending';
      default:
        return 'ascending';
    }
  }
  return direction;
};

export const queryValue = (source, query = '', defaultValue = null) => {
  const value = query.split('.').reduce((result, key) => (result && result[key] ? result[key] : null), source);
  return value || defaultValue;
};

export const FILTER_BY_TYPE = (t) => (item) => item.type && item.type.displayName === t;

export const getProps = (comp, fields = [], defaultValue = null) => {
  const props = comp ? comp.props : null;
  if (!props) return defaultValue;
  return {

    ...props,
    fields,
  };
};

export const extractFields = (items) => {
  const container = items.find(FILTER_BY_TYPE(FIELDS_COMPONENT_TYPE));
  const children = container
    ? React.Children.toArray(container.props.children) : [];
  return children
    .filter(FILTER_BY_TYPE(FIELD_COMPONENT_TYPE))
    .map((c) => c.props);
};

export const extractPagination = (items) => {
  const container = items.find(FILTER_BY_TYPE(PAGINATION_COMPONENT_TYPE));
  return getProps(container, [], {});
};

export const extractQueryFields = (items) => {
  const fields = extractFields(items);
  return fields.filter((f) => f.queryable)
    .map((f) => ({

      ...f,
      value: f.name,
    }));
};

export const extractForms = (items, fields) => ({
  create: getProps(
    items.find(FILTER_BY_TYPE(CREATE_FORM_COMPONENT_TYPE)),
    fields.filter((f) => !f.hideInCreateForm),
  ),
  update: getProps(
    items.find(FILTER_BY_TYPE(UPDATE_FORM_COMPONENT_TYPE)),
    fields.filter((f) => !f.hideInUpdateForm),
  ),
  delete: getProps(items.find(FILTER_BY_TYPE(DELETE_FORM_COMPONENT_TYPE))),
});

export default {
  chevron,
  toggleDirection,
  queryValue,
  FILTER_BY_TYPE,
  extractFields,
  getProps,
  extractForms,
  extractPagination,
  extractQueryFields,
};
