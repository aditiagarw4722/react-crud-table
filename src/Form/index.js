import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Field } from 'formik';
import { Form } from './wrappers';
import Button from '../Button';

const DEFAULT_VALIDATE = () => ({});

const generalValidationError = ({ touched, errors, message }) => {
  const fieldsTouched = touched || {};
  const showErrorMessage = Object.keys(errors || {}).reduce(
    (acc, key) => acc || fieldsTouched[key],
    false
  );
  return (
    showErrorMessage && (
      <Form.ErrorMessage>
        {message || 'There are some errors'}
      </Form.ErrorMessage>
    )
  );
};

const BasicForm = ({ data, onSubmit, initialValues }) => (
  <div>
    <Formik
      enableReinitialize
      initialValues={initialValues}
      validate={data.validate || DEFAULT_VALIDATE}
      onSubmit={onSubmit}
      render={({ errors, touched, error }) => (
        <Form>
          {data.message && <Form.Message>{data.message}</Form.Message>}
          {error && <Form.ErrorMessage>{error}</Form.ErrorMessage>}
          {generalValidationError({
            touched,
            errors,
            message: data.generalErrorMessage,
          })}
          {data.fields.map((field) => (
            <Form.FieldContainer key={field.name}>
              <Form.Label htmlFor={field.name}>{field.label}</Form.Label>
              <Field
                name={field.name}
                placeholder={field.placeholder}
                render={field.render}
                type={field.type}
                readOnly={field.readOnly}
              />
              {errors[field.name] && touched[field.name] && (
                <Form.FieldError>{errors[field.name]}</Form.FieldError>
              )}
            </Form.FieldContainer>
          ))}
          {generalValidationError({
            touched,
            errors,
            message: data.generalErrorMessage,
          })}
          <Button
            type="submit"
            modifiers="positive"
            {...(data.submitButtonProps || {})}
          >
            {data.submitText}
          </Button>
        </Form>
      )}
    />
  </div>
);

BasicForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  data: PropTypes.instanceOf(Object),
  initialValues: PropTypes.instanceOf(Object),
};

BasicForm.defaultProps = {
  initialValues: {},
  data: {
    message: null,
    fields: [],
    submitText: null,
  },
};

export default BasicForm;
