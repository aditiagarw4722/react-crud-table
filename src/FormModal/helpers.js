export const onSubmitHandler = (onSubmit, shouldReset, callback) => (
  values,
  { setError, resetForm, setSubmitting }
) => {
  const reset = Object.keys(values).reduce(
    (result, prop) => ({
      ...result,
      [prop]: '',
    }),
    {}
  );

  const result = onSubmit(values);
  return result
    .then(() => {
      if (shouldReset) {
        resetForm(reset);
      }

      setSubmitting(false);
      callback();
    })
    .catch((err) => {
      setError(err ? err.message : 'Unexpected error');
      setSubmitting(false);
    });
};
