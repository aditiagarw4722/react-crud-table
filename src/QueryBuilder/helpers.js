import React from 'react';
import DateTime from 'react-datetime';
import Label from '../Label';
import { CONDITIONS_LABEL, CONDITIONS } from './constants';

let styles;

const Input = (props) => <input {...props} />;

export const isRuleComplete = (rule) => {
  const { condition, value, field } = rule;
  return condition !== '' && value !== '' && field !== '';
};

export const isBoolean = (type) => type === 'boolean';

export const mapFieldsToOptions = (fields) =>
  fields.map((x) => ({
    text: x.label,
    value: x.value,
    key: x.value,
  }));

export const defaultRuleRender = (rule) => {
  if (isBoolean(rule.type)) {
    return (
      <span>
        <Label style={styles.label}>{rule.condition}</Label>
        &nbsp;
        <Label style={styles.label}>{rule.label}</Label>
      </span>
    );
  }
  return (
    <span>
      <Label style={styles.label}>{rule.label}</Label>
      &nbsp;
      <Label style={styles.label}>{CONDITIONS_LABEL[rule.condition]}</Label>
      &nbsp;
      <Label style={styles.label}>{rule.value}</Label>
    </span>
  );
};

export const conditionsForType = (type) => {
  let result;
  switch (type) {
    case 'number':
    case 'date':
      result = [
        CONDITIONS.EQUALS_TO,
        CONDITIONS.LESS_THAN,
        CONDITIONS.GREATER_THAN,
        CONDITIONS.LESS_OR_EQUALS_THAN,
        CONDITIONS.GREATER_OR_EQUALS_THAN,
        CONDITIONS.IS_NOT_EQUALS_TO,
        CONDITIONS.IS_NOT_LESS_THAN,
        CONDITIONS.IS_NOT_GREATER_THAN,
        CONDITIONS.IS_NOT_LESS_OR_EQUALS_THAN,
        CONDITIONS.IS_NOT_GREATER_OR_EQUALS_THAN,
      ];
      break;
    case 'boolean':
      result = [CONDITIONS.IS, CONDITIONS.IS_NOT];
      break;
    default:
      result = [
        CONDITIONS.CONTAINS,
        CONDITIONS.EQUALS_TO,
        CONDITIONS.BEGINS_WITH,
        CONDITIONS.ENDS_WITH,
        CONDITIONS.IS_NOT_EQUALS_TO,
        CONDITIONS.DOES_NOT_BEGIN_WITH,
        CONDITIONS.DOES_NOT_END_WITH,
        CONDITIONS.DOES_NOT_CONTAIN,
      ];
      break;
  }

  return result.map((r) => ({
    value: r,
    text: CONDITIONS_LABEL[r],
    key: r,
  }));
};

export const getDefaultConditionForType = (type) => {
  switch (type) {
    case 'number':
      return CONDITIONS.EQUALS_TO;
    case 'date':
      return CONDITIONS.GREATER_OR_EQUALS_THAN;
    default:
      return CONDITIONS.CONTAINS;
  }
};

export const inputForType = (type, props) => {
  switch (type) {
    case 'number':
      return <Input type="number" {...props} />;
    case 'boolean':
      return (
        <Input
          {...props}
          type="checkbox"
          onClick={(evt) => {
            props.onChange({
              ...evt,
              target: {
                ...evt.target,
                value: evt.target.checked,
              },
            });
          }}
          onChange={() => {}}
        />
      );
    case 'date':
      return (
        <DateTime
          {...props}
          className="ui input"
          dateFormat="YYYY-MM-DD"
          timeFormat="hh:mm A"
          onChange={(data) => {
            props.onChange({
              target: {
                value: data.format('YYYY-MM-DD hh:mm A'),
              },
            });
          }}
        />
      );
    default:
      return <Input type="text" {...props} />;
  }
};

styles = {
  label: {
    marginBottom: '5px',
  },
};

export default {
  defaultRuleRender,
  conditionsForType,
  getDefaultConditionForType,
  inputForType,
  mapFieldsToOptions,
  isBoolean,
  isRuleComplete,
};
