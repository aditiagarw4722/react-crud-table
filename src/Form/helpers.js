// @flow

import React from 'react';
import { queryValue } from '../CRUDTable/helpers';

import { Form } from './wrappers';

export const DEFAULT_VALIDATE = (): Object => ({});

type Payload = {
  touched: Object,
  errors: Object,
  message: ?string,
};

export const generalValidationError = (
  payload: Payload
): boolean | React$Element<any> => {
  const showErrorMessage = Object.keys(payload.errors).reduce(
    (acc, key) => acc || payload.touched[key],
    false
  );
  return (
    showErrorMessage && (
      <Form.ErrorMessage>
        {queryValue(payload, 'message', 'There are some errors')}
      </Form.ErrorMessage>
    )
  );
};
