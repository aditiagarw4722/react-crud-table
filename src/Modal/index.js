import React from 'react';
import PropTypes from 'prop-types';
import { Container } from './wrappers';
import Button from '../Button';
import { NO_OP } from '../helpers';

const getDisplay = (visible) => (visible ? 'block' : 'none');

const Modal = (props) => {
  const { children, trigger, title, visible, onHide, onShow } = props;
  const style = {
    display: getDisplay(visible),
  };
  return (
    <div>
      {trigger && (
        <Button modifiers="positive" onClick={onShow}>
          {trigger}
        </Button>
      )}
      <Container style={style}>
        <Container.BG onClick={onHide} />
        <Container.Modal>
          {title && <Container.Title>{title}</Container.Title>}
          {children}
        </Container.Modal>
      </Container>
    </div>
  );
};

Modal.propTypes = {
  title: PropTypes.string,
  onHide: PropTypes.func,
  onShow: PropTypes.func,
  children: PropTypes.node,
  trigger: PropTypes.node,
  visible: PropTypes.bool,
};

Modal.defaultProps = {
  onHide: NO_OP,
  onShow: NO_OP,
  title: '',
  children: null,
  trigger: null,
  visible: false,
};

export default Modal;
