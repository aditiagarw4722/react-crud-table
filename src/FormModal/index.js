// @flow

import * as React from 'react';

import Modal from '../Modal';
import Form from '../Form';
import { NO_OP } from '../helpers';
import { onSubmitHandler } from './helpers';
import { scrollToTop } from '../CRUDTable/helpers';

type Props = {
  visible: boolean,
  onVisibilityChange: Function,
  onSubmit: Function,
  shouldReset: boolean,
  trigger: number | string | null | React.Element<any> | Array<any>,
  data: Object,
  initialValues?: Object,
};

type State = {
  key: number,
};

class FormModal extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { key: 0 };
  }

  props: Props;

  render(): React$Element<any> {
    const {
      data,
      trigger,
      initialValues,
      shouldReset,
      onSubmit,
      visible,
      onVisibilityChange,
    } = this.props;
    const { key } = this.state;
    return (
      <Modal
        trigger={trigger}
        title={data.title}
        visible={visible}
        onShow={() => {
          scrollToTop();
          onVisibilityChange(true);
          this.setState({
            key: new Date().getTime(),
          });
        }}
        onHide={() => {
          onVisibilityChange(false);
        }}
      >
        <Form
          key={key}
          data={data}
          initialValues={initialValues}
          onSubmit={onSubmitHandler(onSubmit, shouldReset, () =>
            onVisibilityChange(false)
          )}
        />
      </Modal>
    );
  }
}

// $FlowFixMe
FormModal.defaultProps = {
  onVisibilityChange: NO_OP,
  onSubmit: NO_OP,
  shouldReset: false,
  trigger: null,
  initialValues: null,
  visible: false,
};

export default FormModal;
