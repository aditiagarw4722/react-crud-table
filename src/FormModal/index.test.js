import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';

import FormModal from './';
import Modal from '../Modal';
import Form from '../Form';

jest.mock('./helpers');

import { onSubmitHandler } from './helpers';

const renderer = new ShallowRenderer();

describe('FormFormModal', () => {
  let props;

  beforeEach(() => {
    props = {
      trigger: 'Open',
      data: {
        title: 'Test title',
        fields: [],
      },
      onVisibilityChange: jest.fn(),
      onSubmit: jest.fn(() => Promise.resolve({ success: true })),
      shouldReset: false,
      visible: true,
      initialValues: { name: 'John' },
    };

    onSubmitHandler.mockClear().mockImplementation((a, b, c) => {
      return (v) => {
        a(v);
        c();
      };
    });
  });

  it('renders as expected with the given props', () => {
    const result = renderer.render(<FormModal {...props} />);
    expect(result).toMatchSnapshot();
  });

  it('notifies when opened', () => {
    const result = create(<FormModal {...props} />);

    result.root.findByType(Modal).props.onShow();
    expect(props.onVisibilityChange).toHaveBeenCalledTimes(1);
    expect(props.onVisibilityChange).toHaveBeenCalledWith(true);
  });

  it('notifies when opened', () => {
    const result = create(<FormModal {...props} />);

    result.root.findByType(Modal).props.onHide();
    expect(props.onVisibilityChange).toHaveBeenCalledTimes(1);
    expect(props.onVisibilityChange).toHaveBeenCalledWith(false);
  });

  it('notifies on Submit', () => {
    const result = create(<FormModal {...props} />);

    const values = { name: 'John' };
    const payload = {
      setError: jest.fn(),
      resetForm: jest.fn(),
      setSubmitting: jest.fn(),
    };

    result.root.findByType(Form).props.onSubmit(values, payload);

    expect(props.onSubmit).toHaveBeenCalledTimes(1);
    expect(props.onSubmit).toHaveBeenCalledWith(values);
    expect(props.onVisibilityChange).toHaveBeenCalledTimes(1);
    expect(props.onVisibilityChange).toHaveBeenCalledWith(false);
  });
});
