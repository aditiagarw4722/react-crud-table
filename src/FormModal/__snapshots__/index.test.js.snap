// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFormModal renders as expected with the given props 1`] = `
<Modal
  onHide={[Function]}
  onShow={[Function]}
  title="Test title"
  trigger="Open"
  visible={true}
>
  <BasicForm
    data={
      Object {
        "fields": Array [],
        "title": "Test title",
      }
    }
    initialValues={
      Object {
        "name": "John",
      }
    }
    onSubmit={[Function]}
  />
</Modal>
`;
